File: runners/java/NegativeV1.java
// Java runner for the rfc9421_negative_v1 CORE cross-language battery.
//
// Self-contained probe: unlike the python/typescript/go/rust runners (which import
// the published AlgoVoi verifier for their language), no Java RFC 9421 verifier
// exists yet, so this runner re-implements the exact verdict surface inline,
// ported from the reference modules (signing_base.py, parse.py, keycheck.py,
// verify.py) and the algovoi-rfc9421-ecdsa provider. It must produce byte-identical
// verdicts to the other runners on every case: reject every negative, accept every
// positive control.
//
// Crypto: Ed25519 verify + canonical-S enforcement via Bouncy Castle
// (Ed25519Signer rejects non-canonical S and wrong-length signatures); the
// small-order / non-canonical public-key gate is ported from keycheck.py with
// java.math.BigInteger (RFC 8032 Appendix A arithmetic), applied fail-closed
// BEFORE verification. ECDSA P-256/P-384 uses the JDK provider over a raw r||s
// signature with explicit [1,n-1] range, fixed-width, on-curve and strict-low-s
// checks.
//
// Corpus path: argv[0], else $ALGOVOI_NEGATIVE_V1, else the sibling repo default.

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1Integer;
import org.bouncycastle.asn1.DERSequence;
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters;
import org.bouncycastle.crypto.signers.Ed25519Signer;

import java.io.File;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.AlgorithmParameters;
import java.security.KeyFactory;
import java.security.Signature;
import java.security.spec.ECGenParameterSpec;
import java.security.spec.ECParameterSpec;
import java.security.spec.ECPoint;
import java.security.spec.ECPublicKeySpec;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class NegativeV1 {

    static final String DEFAULT_PATH =
        "../../corpus/rfc9421_negative_v1/rfc9421_negative_v1.json";

    // ---- errors that mirror the reference's fail-closed exceptions ----
    static final class SigningBaseError extends RuntimeException {
        SigningBaseError(String m) { super(m); }
    }
    static final class ParseError extends RuntimeException {
        ParseError(String m) { super(m); }
    }
    static final class WeakKeyError extends RuntimeException {
        WeakKeyError(String m) { super(m); }
    }

    // ================= signing base (port of signing_base.py) =================
    static String buildSigningBase(JsonNode in, String mode, String signatureParamsRaw) {
        if (!mode.equals("algovoi-v0") && !mode.equals("rfc9421")) {
            throw new SigningBaseError("mode must be 'algovoi-v0' or 'rfc9421', got " + mode);
        }
        if (mode.equals("rfc9421") && signatureParamsRaw == null) {
            throw new SigningBaseError("rfc9421 mode requires signature_params_raw");
        }
        Map<String, String> headers = new LinkedHashMap<>();
        JsonNode h = in.get("headers");
        if (h != null && h.isObject()) {
            for (Iterator<String> it = h.fieldNames(); it.hasNext(); ) {
                String k = it.next();
                headers.put(k.toLowerCase(), h.get(k).asText());
            }
        }
        JsonNode params = in.get("parameters");

        List<String> lines = new ArrayList<>();
        JsonNode covered = in.get("covered_components");
        for (JsonNode compNode : covered) {
            String component = compNode.asText();
            String c = component.toLowerCase();
            String value;
            switch (c) {
                case "@method":
                    value = strOrThrow(in, "method", "@method covered but method not supplied");
                    if (!mode.equals("rfc9421")) value = value.toLowerCase();
                    break;
                case "@authority":
                    value = strOrThrow(in, "authority", "@authority covered but authority not supplied").toLowerCase();
                    break;
                case "@path":
                    value = strOrThrow(in, "path", "@path covered but path not supplied");
                    break;
                case "@target-uri":
                    value = strOrThrow(in, "target_uri", "@target-uri covered but target_uri not supplied");
                    break;
                case "@scheme":
                    value = strOrThrow(in, "scheme", "@scheme covered but scheme not supplied").toLowerCase();
                    break;
                case "@status":
                    if (in.get("status") == null || in.get("status").isNull())
                        throw new SigningBaseError("@status covered but status not supplied");
                    value = in.get("status").bigIntegerValue().toString();
                    break;
                case "created":
                    value = paramStrOrThrow(params, "created",
                        "'created' covered but no 'created' parameter in Signature-Input");
                    break;
                case "expires":
                    value = paramStrOrThrow(params, "expires",
                        "'expires' covered but no 'expires' parameter in Signature-Input");
                    break;
                default:
                    if (c.startsWith("@"))
                        throw new SigningBaseError("unsupported derived component: " + component);
                    if (!headers.containsKey(c))
                        throw new SigningBaseError("covered header " + component + " not present in supplied headers");
                    value = headers.get(c);
            }
            lines.add("\"" + c + "\": " + value);
        }
        if (mode.equals("rfc9421")) {
            lines.add("\"@signature-params\": " + signatureParamsRaw);
        }
        return String.join("\n", lines);
    }

    static String strOrThrow(JsonNode in, String field, String msg) {
        JsonNode n = in.get(field);
        if (n == null || n.isNull()) throw new SigningBaseError(msg);
        return n.asText();
    }

    static String paramStrOrThrow(JsonNode params, String key, String msg) {
        if (params == null || !params.has(key)) throw new SigningBaseError(msg);
        JsonNode v = params.get(key);
        return v.isIntegralNumber() ? v.bigIntegerValue().toString() : v.asText();
    }

    // ================= structured-field parse (port of parse.py) =================
    static void parseSignatureInput(String header) {
        if (header == null) throw new ParseError("header must be str");
        String hv = header.strip();
        if (hv.isEmpty()) throw new ParseError("empty header value");
        String rest;
        java.util.regex.Matcher lm =
            java.util.regex.Pattern.compile("^\\s*([A-Za-z][A-Za-z0-9_-]*)\\s*=\\s*").matcher(hv);
        if (lm.find()) {
            rest = hv.substring(lm.end());
        } else if (hv.startsWith("(")) {
            rest = hv;
        } else {
            throw new ParseError("no label or covered-components list found at start");
        }
        java.util.regex.Matcher cm =
            java.util.regex.Pattern.compile("\\(\\s*(?:\"[^\"]*\"\\s*)*\\)").matcher(rest);
        if (!cm.lookingAt()) throw new ParseError("no covered-components list found");
        // parameters after the covered list are optional and, per parse.py, are
        // tolerantly scanned; only the structural pieces above can reject.
    }

    static void parseSignatureValue(String header) {
        if (header == null) throw new ParseError("header must be str");
        String hv = header.strip();
        if (hv.isEmpty()) throw new ParseError("empty Signature header value");
        String rest;
        java.util.regex.Matcher lm =
            java.util.regex.Pattern.compile("^\\s*([A-Za-z][A-Za-z0-9_-]*)\\s*=\\s*").matcher(hv);
        if (lm.find()) {
            rest = hv.substring(lm.end()).strip();
        } else if (hv.startsWith(":")) {
            rest = hv;
        } else {
            throw new ParseError("no label or signature-value prefix found at start");
        }
        if (!rest.startsWith(":") || !rest.endsWith(":") || rest.length() < 2) {
            throw new ParseError("signature value must be wrapped in colons");
        }
        String sigB64 = rest.substring(1, rest.length() - 1);
        byte[] decoded;
        try {
            decoded = Base64.getDecoder().decode(sigB64);
        } catch (IllegalArgumentException e) {
            throw new ParseError("signature value is not valid base64");
        }
        // Reject NON-CANONICAL base64 (non-zero pad bits): require exact round-trip.
        if (!Base64.getEncoder().encodeToString(decoded).equals(sigB64)) {
            throw new ParseError("signature value is not canonical base64 (non-zero pad bits)");
        }
    }

    // ================= Ed25519 key gate (port of keycheck.py) =================
    static final BigInteger P = BigInteger.TWO.pow(255).subtract(BigInteger.valueOf(19));
    static final BigInteger D = BigInteger.valueOf(-121665)
        .multiply(BigInteger.valueOf(121666).modInverse(P)).mod(P);
    static final BigInteger SQRT_M1 =
        BigInteger.TWO.modPow(P.subtract(BigInteger.ONE).divide(BigInteger.valueOf(4)), P);

    static BigInteger recoverX(BigInteger y, int sign) {
        if (y.compareTo(P) >= 0) return null;
        BigInteger y2 = y.multiply(y).mod(P);
        BigInteger num = y2.subtract(BigInteger.ONE).mod(P);
        BigInteger den = D.multiply(y2).add(BigInteger.ONE).mod(P);
        BigInteger x2 = num.multiply(den.modInverse(P)).mod(P);
        if (x2.signum() == 0) return (sign == 1) ? null : BigInteger.ZERO;
        BigInteger x = x2.modPow(P.add(BigInteger.valueOf(3)).divide(BigInteger.valueOf(8)), P);
        if (x.multiply(x).subtract(x2).mod(P).signum() != 0) x = x.multiply(SQRT_M1).mod(P);
        if (x.multiply(x).subtract(x2).mod(P).signum() != 0) return null;
        if (x.testBit(0) ? sign == 0 : sign == 1) x = P.subtract(x);
        return x;
    }

    // extended coords (X, Y, Z, T)
    static BigInteger[] decodePoint(byte[] pk) {
        if (pk.length != 32) return null;
        byte[] le = new byte[32];
        for (int i = 0; i < 32; i++) le[i] = pk[31 - i];
        BigInteger y = new BigInteger(1, le);
        int sign = y.testBit(255) ? 1 : 0;
        y = y.clearBit(255);
        BigInteger x = recoverX(y, sign);
        if (x == null) return null;
        return new BigInteger[]{x, y, BigInteger.ONE, x.multiply(y).mod(P)};
    }

    static BigInteger[] add(BigInteger[] p, BigInteger[] q) {
        BigInteger a = p[1].subtract(p[0]).multiply(q[1].subtract(q[0])).mod(P);
        BigInteger b = p[1].add(p[0]).multiply(q[1].add(q[0])).mod(P);
        BigInteger c = BigInteger.TWO.multiply(p[3]).multiply(q[3]).multiply(D).mod(P);
        BigInteger dd = BigInteger.TWO.multiply(p[2]).multiply(q[2]).mod(P);
        BigInteger e = b.subtract(a), f = dd.subtract(c), g = dd.add(c), hh = b.add(a);
        return new BigInteger[]{e.multiply(f).mod(P), g.multiply(hh).mod(P),
                                f.multiply(g).mod(P), e.multiply(hh).mod(P)};
    }

    static BigInteger[] mul8(BigInteger[] p) {
        BigInteger[] p2 = add(p, p);
        BigInteger[] p4 = add(p2, p2);
        return add(p4, p4);
    }

    static boolean isIdentity(BigInteger[] p) {
        return p[0].mod(P).signum() == 0 && p[1].subtract(p[2]).mod(P).signum() == 0;
    }

    static boolean isSmallOrder(byte[] pk) {
        BigInteger[] p = decodePoint(pk);
        if (p == null) return false;
        return isIdentity(mul8(p));
    }

    static void checkEd25519PublicKey(byte[] pk) {
        BigInteger[] p = decodePoint(pk);
        if (p == null) throw new WeakKeyError("public key is not a canonical on-curve Ed25519 point");
        if (isIdentity(mul8(p))) throw new WeakKeyError("public key is a small-order Ed25519 point");
    }

    // ================= Ed25519 verify (port of verify.py) =================
    static boolean ed25519Verify(byte[] msg, byte[] sig, byte[] pk) {
        if (sig.length != 64) return false;      // reference: wrong length -> invalid
        if (pk.length != 32) return false;
        try {
            checkEd25519PublicKey(pk);            // fail-closed key gate before verify
        } catch (WeakKeyError e) {
            return false;
        }
        Ed25519PublicKeyParameters pub = new Ed25519PublicKeyParameters(pk, 0);
        Ed25519Signer signer = new Ed25519Signer();
        signer.init(false, pub);
        signer.update(msg, 0, msg.length);
        return signer.verifySignature(sig);       // BC rejects non-canonical S
    }

    // ================= ECDSA verify (port of algovoi-rfc9421-ecdsa) =================
    static boolean ecdsaVerify(String curve, byte[] msg, byte[] sigRaw, byte[] pub, boolean strictLowS) {
        try {
            int fieldSize = curve.equals("p256") ? 32 : 48;
            String stdName = curve.equals("p256") ? "secp256r1" : "secp384r1";
            String sigAlg = curve.equals("p256") ? "SHA256withECDSA" : "SHA384withECDSA";
            if (sigRaw.length != 2 * fieldSize) return false;   // wrong width

            BigInteger r = new BigInteger(1, java.util.Arrays.copyOfRange(sigRaw, 0, fieldSize));
            BigInteger s = new BigInteger(1, java.util.Arrays.copyOfRange(sigRaw, fieldSize, 2 * fieldSize));

            AlgorithmParameters ap = AlgorithmParameters.getInstance("EC");
            ap.init(new ECGenParameterSpec(stdName));
            ECParameterSpec spec = ap.getParameterSpec(ECParameterSpec.class);
            BigInteger n = spec.getOrder();

            // r, s in [1, n-1]
            if (r.signum() <= 0 || r.compareTo(n) >= 0 || s.signum() <= 0 || s.compareTo(n) >= 0)
                return false;
            if (strictLowS && s.compareTo(n.shiftRight(1)) > 0) return false;

            // decode SEC1 uncompressed public point and validate on-curve
            if (pub.length != 1 + 2 * fieldSize || pub[0] != 0x04) return false;
            BigInteger x = new BigInteger(1, java.util.Arrays.copyOfRange(pub, 1, 1 + fieldSize));
            BigInteger yc = new BigInteger(1, java.util.Arrays.copyOfRange(pub, 1 + fieldSize, 1 + 2 * fieldSize));
            BigInteger pMod = ((java.security.spec.ECFieldFp) spec.getCurve().getField()).getP();
            BigInteger aa = spec.getCurve().getA(), bb = spec.getCurve().getB();
            BigInteger lhs = yc.multiply(yc).mod(pMod);
            BigInteger rhs = x.multiply(x).multiply(x).add(aa.multiply(x)).add(bb).mod(pMod);
            if (!lhs.equals(rhs)) return false;      // off-curve public key

            ECPublicKeySpec pubSpec = new ECPublicKeySpec(new ECPoint(x, yc), spec);
            java.security.PublicKey key = KeyFactory.getInstance("EC").generatePublic(pubSpec);

            // DER-encode r||s for the JDK verifier
            ASN1EncodableVector v = new ASN1EncodableVector();
            v.add(new ASN1Integer(r));
            v.add(new ASN1Integer(s));
            byte[] der = new DERSequence(v).getEncoded("DER");

            Signature verifier = Signature.getInstance(sigAlg);
            verifier.initVerify(key);
            verifier.update(msg);
            return verifier.verify(der);
        } catch (Exception e) {
            return false;   // any setup failure -> invalid, matching the reference
        }
    }

    // ================= driver =================
    static byte[] hex(String s) {
        int n = s.length();
        byte[] out = new byte[n / 2];
        for (int i = 0; i < n; i += 2) out[i / 2] = (byte) Integer.parseInt(s.substring(i, i + 2), 16);
        return out;
    }

    public static void main(String[] args) throws Exception {
        String path = args.length > 0 ? args[0]
            : System.getenv().getOrDefault("ALGOVOI_NEGATIVE_V1", DEFAULT_PATH);
        JsonNode corpus = new ObjectMapper().readTree(new File(path));

        int total = 0, matched = 0;
        List<String> fails = new ArrayList<>();

        for (JsonNode c : corpus.get("signing_base")) {
            total++;
            boolean ok;
            String want = new String(Base64.getDecoder().decode(c.get("signing_base_b64").asText()), StandardCharsets.UTF_8);
            String spr = c.hasNonNull("signature_params_raw") ? c.get("signature_params_raw").asText() : null;
            try {
                // build FIRST; `ok && build(...)` would short-circuit for negative
                // cases and never verify that the build actually fails.
                String built = buildSigningBase(c.get("in"), c.get("mode").asText(), spr);
                ok = c.get("ok").asBoolean() && built.equals(want);
            } catch (RuntimeException e) {
                ok = !c.get("ok").asBoolean();
            }
            record(ok, "signing_base", c, fails); if (ok) matched++;
        }
        for (JsonNode c : corpus.get("signature_input_parse")) {
            total++;
            boolean parsedOk;
            try { parseSignatureInput(c.get("header").asText()); parsedOk = true; }
            catch (ParseError e) { parsedOk = false; }
            boolean ok = parsedOk == c.get("ok").asBoolean();
            record(ok, "sig_input_parse", c, fails); if (ok) matched++;
        }
        for (JsonNode c : corpus.get("signature_value_parse")) {
            total++;
            boolean parsedOk;
            try { parseSignatureValue(c.get("header").asText()); parsedOk = true; }
            catch (ParseError e) { parsedOk = false; }
            boolean ok = parsedOk == c.get("ok").asBoolean();
            record(ok, "sig_value_parse", c, fails); if (ok) matched++;
        }
        for (JsonNode c : corpus.get("keygate")) {
            total++;
            byte[] raw = hex(c.get("pk_hex").asText());
            String rejected;
            try { checkEd25519PublicKey(raw); rejected = null; }
            catch (WeakKeyError e) { rejected = "WeakKeyError"; }
            String want = c.hasNonNull("rejected") ? c.get("rejected").asText() : null;
            boolean ok = java.util.Objects.equals(rejected, want)
                && isSmallOrder(raw) == c.get("small_order").asBoolean();
            record(ok, "keygate", c, fails); if (ok) matched++;
        }
        for (JsonNode c : corpus.get("ed25519_verify")) {
            total++;
            byte[] base = Base64.getDecoder().decode(c.get("signing_base_b64").asText());
            boolean valid = ed25519Verify(base, hex(c.get("sig_hex").asText()), hex(c.get("pk_hex").asText()));
            boolean ok = valid == c.get("expect_valid").asBoolean();
            record(ok, "ed25519_verify", c, fails); if (ok) matched++;
        }
        for (JsonNode c : corpus.get("ecdsa_verify")) {
            total++;
            byte[] msg = hex(c.get("msg_hex").asText());
            boolean strict = c.hasNonNull("strict_low_s") && c.get("strict_low_s").asBoolean();
            boolean valid = ecdsaVerify(c.get("curve").asText(), msg,
                hex(c.get("sig_raw_hex").asText()), hex(c.get("pub_uncompressed_hex").asText()), strict);
            boolean ok = valid == c.get("expect_valid").asBoolean();
            record(ok, "ecdsa_verify", c, fails); if (ok) matched++;
        }

        for (String f : fails) System.out.println("FAIL  " + f);
        System.out.println("\njava: " + matched + "/" + total + " cases matched");
        System.exit(matched == total ? 0 : 1);
    }

    static void record(boolean ok, String section, JsonNode c, List<String> fails) {
        if (!ok) {
            String note = c.hasNonNull("note") ? c.get("note").asText() : "";
            fails.add("[" + section + "] " + note);
        }
    }
}

File: runners/kotlin/NegativeV1.kt
// Kotlin runner for the rfc9421_negative_v1 CORE cross-language battery.
//
// Independent Kotlin/JVM port of the verdict surface (signing_base.py, parse.py,
// keycheck.py, verify.py + the algovoi-rfc9421-ecdsa provider). Like the Java
// runner it is self-contained (no Kotlin RFC 9421 verifier exists) and must
// produce byte-identical verdicts to every other runner. Ed25519 verify +
// canonical-S via Bouncy Castle; the small-order / non-canonical key gate is
// ported with java.math.BigInteger; ECDSA P-256/P-384 via the JDK provider with
// explicit range / width / on-curve / strict-low-s checks.
//
// Corpus path: argv[0], else $ALGOVOI_NEGATIVE_V1, else the sibling repo default.

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.bouncycastle.asn1.ASN1EncodableVector
import org.bouncycastle.asn1.ASN1Integer
import org.bouncycastle.asn1.DERSequence
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import java.io.File
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.Signature
import java.security.spec.ECFieldFp
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec
import java.util.Base64
import kotlin.system.exitProcess

private const val DEFAULT_PATH = "../../corpus/rfc9421_negative_v1/rfc9421_negative_v1.json"

private class SigningBaseError(m: String) : RuntimeException(m)
private class ParseError(m: String) : RuntimeException(m)
private class WeakKeyError(m: String) : RuntimeException(m)

// ================= signing base (port of signing_base.py) =================
private fun buildSigningBase(input: JsonNode, mode: String, sigParamsRaw: String?): String {
    if (mode != "algovoi-v0" && mode != "rfc9421")
        throw SigningBaseError("mode must be 'algovoi-v0' or 'rfc9421', got $mode")
    if (mode == "rfc9421" && sigParamsRaw == null)
        throw SigningBaseError("rfc9421 mode requires signature_params_raw")

    val headers = LinkedHashMap<String, String>()
    input.get("headers")?.takeIf { it.isObject }?.let { h ->
        h.fieldNames().forEach { k -> headers[k.lowercase()] = h.get(k).asText() }
    }
    val params = input.get("parameters")

    fun strOrThrow(field: String, msg: String): String {
        val n = input.get(field)
        if (n == null || n.isNull) throw SigningBaseError(msg)
        return n.asText()
    }
    fun paramStrOrThrow(key: String, msg: String): String {
        if (params == null || !params.has(key)) throw SigningBaseError(msg)
        val v = params.get(key)
        return if (v.isIntegralNumber) v.bigIntegerValue().toString() else v.asText()
    }

    val lines = ArrayList<String>()
    for (compNode in input.get("covered_components")) {
        val component = compNode.asText()
        val c = component.lowercase()
        val value: String = when (c) {
            "@method" -> strOrThrow("method", "@method covered but method not supplied")
                .let { if (mode == "rfc9421") it else it.lowercase() }
            "@authority" -> strOrThrow("authority", "@authority covered but authority not supplied").lowercase()
            "@path" -> strOrThrow("path", "@path covered but path not supplied")
            "@target-uri" -> strOrThrow("target_uri", "@target-uri covered but target_uri not supplied")
            "@scheme" -> strOrThrow("scheme", "@scheme covered but scheme not supplied").lowercase()
            "@status" -> {
                val n = input.get("status")
                if (n == null || n.isNull) throw SigningBaseError("@status covered but status not supplied")
                n.bigIntegerValue().toString()
            }
            "created" -> paramStrOrThrow("created", "'created' covered but no 'created' parameter in Signature-Input")
            "expires" -> paramStrOrThrow("expires", "'expires' covered but no 'expires' parameter in Signature-Input")
            else -> {
                if (c.startsWith("@")) throw SigningBaseError("unsupported derived component: $component")
                headers[c] ?: throw SigningBaseError("covered header $component not present in supplied headers")
            }
        }
        lines.add("\"$c\": $value")
    }
    if (mode == "rfc9421") lines.add("\"@signature-params\": $sigParamsRaw")
    return lines.joinToString("\n")
}

// ================= structured-field parse (port of parse.py) =================
private val LABEL_RE = Regex("^\\s*([A-Za-z][A-Za-z0-9_-]*)\\s*=\\s*")
private val COVERED_RE = Regex("\\(\\s*(?:\"[^\"]*\"\\s*)*\\)")

private fun parseSignatureInput(header: String?) {
    if (header == null) throw ParseError("header must be str")
    val hv = header.trim()
    if (hv.isEmpty()) throw ParseError("empty header value")
    val lm = LABEL_RE.find(hv)
    val rest = when {
        lm != null -> hv.substring(lm.range.last + 1)
        hv.startsWith("(") -> hv
        else -> throw ParseError("no label or covered-components list found at start")
    }
    val cm = COVERED_RE.find(rest)
    if (cm == null || cm.range.first != 0) throw ParseError("no covered-components list found")
}

private fun parseSignatureValue(header: String?) {
    if (header == null) throw ParseError("header must be str")
    val hv = header.trim()
    if (hv.isEmpty()) throw ParseError("empty Signature header value")
    val lm = LABEL_RE.find(hv)
    val rest = when {
        lm != null -> hv.substring(lm.range.last + 1).trim()
        hv.startsWith(":") -> hv
        else -> throw ParseError("no label or signature-value prefix found at start")
    }
    if (!rest.startsWith(":") || !rest.endsWith(":") || rest.length < 2)
        throw ParseError("signature value must be wrapped in colons")
    val sigB64 = rest.substring(1, rest.length - 1)
    val decoded = try { Base64.getDecoder().decode(sigB64) }
        catch (e: IllegalArgumentException) { throw ParseError("signature value is not valid base64") }
    if (Base64.getEncoder().encodeToString(decoded) != sigB64)
        throw ParseError("signature value is not canonical base64 (non-zero pad bits)")
}

// ================= Ed25519 key gate (port of keycheck.py) =================
private val P: BigInteger = BigInteger.TWO.pow(255).subtract(BigInteger.valueOf(19))
private val D: BigInteger = BigInteger.valueOf(-121665)
    .multiply(BigInteger.valueOf(121666).modInverse(P)).mod(P)
private val SQRT_M1: BigInteger = BigInteger.TWO.modPow(P.subtract(BigInteger.ONE).divide(BigInteger.valueOf(4)), P)

private fun recoverX(y: BigInteger, sign: Int): BigInteger? {
    if (y >= P) return null
    val y2 = y.multiply(y).mod(P)
    val x2 = y2.subtract(BigInteger.ONE).mod(P)
        .multiply(D.multiply(y2).add(BigInteger.ONE).mod(P).modInverse(P)).mod(P)
    if (x2.signum() == 0) return if (sign == 1) null else BigInteger.ZERO
    var x = x2.modPow(P.add(BigInteger.valueOf(3)).divide(BigInteger.valueOf(8)), P)
    if (x.multiply(x).subtract(x2).mod(P).signum() != 0) x = x.multiply(SQRT_M1).mod(P)
    if (x.multiply(x).subtract(x2).mod(P).signum() != 0) return null
    if (if (x.testBit(0)) sign == 0 else sign == 1) x = P.subtract(x)
    return x
}

private fun decodePoint(pk: ByteArray): Array<BigInteger>? {
    if (pk.size != 32) return null
    val le = ByteArray(32) { pk[31 - it] }
    var y = BigInteger(1, le)
    val sign = if (y.testBit(255)) 1 else 0
    y = y.clearBit(255)
    val x = recoverX(y, sign) ?: return null
    return arrayOf(x, y, BigInteger.ONE, x.multiply(y).mod(P))
}

private fun add(p: Array<BigInteger>, q: Array<BigInteger>): Array<BigInteger> {
    val a = p[1].subtract(p[0]).multiply(q[1].subtract(q[0])).mod(P)
    val b = p[1].add(p[0]).multiply(q[1].add(q[0])).mod(P)
    val c = BigInteger.TWO.multiply(p[3]).multiply(q[3]).multiply(D).mod(P)
    val dd = BigInteger.TWO.multiply(p[2]).multiply(q[2]).mod(P)
    val e = b.subtract(a); val f = dd.subtract(c); val g = dd.add(c); val h = b.add(a)
    return arrayOf(e.multiply(f).mod(P), g.multiply(h).mod(P), f.multiply(g).mod(P), e.multiply(h).mod(P))
}

private fun mul8(p: Array<BigInteger>): Array<BigInteger> {
    val p2 = add(p, p); val p4 = add(p2, p2); return add(p4, p4)
}

private fun isIdentity(p: Array<BigInteger>) =
    p[0].mod(P).signum() == 0 && p[1].subtract(p[2]).mod(P).signum() == 0

private fun isSmallOrder(pk: ByteArray): Boolean {
    val p = decodePoint(pk) ?: return false
    return isIdentity(mul8(p))
}

private fun checkEd25519PublicKey(pk: ByteArray) {
    val p = decodePoint(pk) ?: throw WeakKeyError("public key is not a canonical on-curve Ed25519 point")
    if (isIdentity(mul8(p))) throw WeakKeyError("public key is a small-order Ed25519 point")
}

// ================= Ed25519 verify (port of verify.py) =================
private fun ed25519Verify(msg: ByteArray, sig: ByteArray, pk: ByteArray): Boolean {
    if (sig.size != 64) return false
    if (pk.size != 32) return false
    try { checkEd25519PublicKey(pk) } catch (e: WeakKeyError) { return false }
    val signer = Ed25519Signer()
    signer.init(false, Ed25519PublicKeyParameters(pk, 0))
    signer.update(msg, 0, msg.size)
    return signer.verifySignature(sig)
}

// ================= ECDSA verify (port of algovoi-rfc9421-ecdsa) =================
private fun ecdsaVerify(curve: String, msg: ByteArray, sigRaw: ByteArray, pub: ByteArray, strictLowS: Boolean): Boolean {
    return try {
        val fieldSize = if (curve == "p256") 32 else 48
        val stdName = if (curve == "p256") "secp256r1" else "secp384r1"
        val sigAlg = if (curve == "p256") "SHA256withECDSA" else "SHA384withECDSA"
        if (sigRaw.size != 2 * fieldSize) return false

        val r = BigInteger(1, sigRaw.copyOfRange(0, fieldSize))
        val s = BigInteger(1, sigRaw.copyOfRange(fieldSize, 2 * fieldSize))

        val ap = AlgorithmParameters.getInstance("EC")
        ap.init(ECGenParameterSpec(stdName))
        val spec = ap.getParameterSpec(ECParameterSpec::class.java)
        val n = spec.order
        if (r.signum() <= 0 || r >= n || s.signum() <= 0 || s >= n) return false
        if (strictLowS && s > n.shiftRight(1)) return false

        if (pub.size != 1 + 2 * fieldSize || pub[0].toInt() != 0x04) return false
        val x = BigInteger(1, pub.copyOfRange(1, 1 + fieldSize))
        val yc = BigInteger(1, pub.copyOfRange(1 + fieldSize, 1 + 2 * fieldSize))
        val pMod = (spec.curve.field as ECFieldFp).p
        val lhs = yc.multiply(yc).mod(pMod)
        val rhs = x.multiply(x).multiply(x).add(spec.curve.a.multiply(x)).add(spec.curve.b).mod(pMod)
        if (lhs != rhs) return false

        val key = KeyFactory.getInstance("EC").generatePublic(ECPublicKeySpec(ECPoint(x, yc), spec))
        val v = ASN1EncodableVector()
        v.add(ASN1Integer(r)); v.add(ASN1Integer(s))
        val der = DERSequence(v).getEncoded("DER")

        val verifier = Signature.getInstance(sigAlg)
        verifier.initVerify(key)
        verifier.update(msg)
        verifier.verify(der)
    } catch (e: Exception) {
        false
    }
}

private fun hex(s: String): ByteArray {
    val out = ByteArray(s.length / 2)
    for (i in s.indices step 2) out[i / 2] = s.substring(i, i + 2).toInt(16).toByte()
    return out
}

fun main(args: Array<String>) {
    val path = if (args.isNotEmpty()) args[0]
        else System.getenv("ALGOVOI_NEGATIVE_V1") ?: DEFAULT_PATH
    val corpus = ObjectMapper().readTree(File(path))
    var total = 0; var matched = 0
    val fails = ArrayList<String>()

    fun record(ok: Boolean, section: String, c: JsonNode) {
        total++
        if (ok) matched++ else fails.add("[$section] " + (if (c.hasNonNull("note")) c.get("note").asText() else ""))
    }

    for (c in corpus.get("signing_base")) {
        val want = String(Base64.getDecoder().decode(c.get("signing_base_b64").asText()), Charsets.UTF_8)
        val spr = if (c.hasNonNull("signature_params_raw")) c.get("signature_params_raw").asText() else null
        val ok = try {
            c.get("ok").asBoolean() && buildSigningBase(c.get("in"), c.get("mode").asText(), spr) == want
        } catch (e: RuntimeException) { !c.get("ok").asBoolean() }
        record(ok, "signing_base", c)
    }
    for (c in corpus.get("signature_input_parse")) {
        val parsedOk = try { parseSignatureInput(c.get("header").asText()); true } catch (e: ParseError) { false }
        record(parsedOk == c.get("ok").asBoolean(), "sig_input_parse", c)
    }
    for (c in corpus.get("signature_value_parse")) {
        val parsedOk = try { parseSignatureValue(c.get("header").asText()); true } catch (e: ParseError) { false }
        record(parsedOk == c.get("ok").asBoolean(), "sig_value_parse", c)
    }
    for (c in corpus.get("keygate")) {
        val raw = hex(c.get("pk_hex").asText())
        val rejected = try { checkEd25519PublicKey(raw); null } catch (e: WeakKeyError) { "WeakKeyError" }
        val want = if (c.hasNonNull("rejected")) c.get("rejected").asText() else null
        record(rejected == want && isSmallOrder(raw) == c.get("small_order").asBoolean(), "keygate", c)
    }
    for (c in corpus.get("ed25519_verify")) {
        val base = Base64.getDecoder().decode(c.get("signing_base_b64").asText())
        val valid = ed25519Verify(base, hex(c.get("sig_hex").asText()), hex(c.get("pk_hex").asText()))
        record(valid == c.get("expect_valid").asBoolean(), "ed25519_verify", c)
    }
    for (c in corpus.get("ecdsa_verify")) {
        val strict = c.hasNonNull("strict_low_s") && c.get("strict_low_s").asBoolean()
        val valid = ecdsaVerify(c.get("curve").asText(), hex(c.get("msg_hex").asText()),
            hex(c.get("sig_raw_hex").asText()), hex(c.get("pub_uncompressed_hex").asText()), strict)
        record(valid == c.get("expect_valid").asBoolean(), "ecdsa_verify", c)
    }

    fails.forEach { println("FAIL  $it") }
    println("\nkotlin: $matched/$total cases matched")
    exitProcess(if (matched == total) 0 else 1)
}
